### PR TITLE
feat: linting the start of declarations and commands

### PR DESCRIPTION
**Pull request by adomani**
This linter checks that
* each command starts at the beginning of a line;
* the "hypotheses segment" of each declaration coincides with its pretty-printed version.

This is likely to have still false positives, but maybe it is acceptable!

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: MathlibTest/PPRoundtrip.lean [hunk 1]
```diff
@@ -64,3 +65,21 @@ set_option linter.ppRoundtrip true in
...
+example {a: Nat} : a = a := rfl
+
+/--
+warning: Current syntax:  ' {a :Nat} '
+Expected syntax: ' {a : Nat}'
+
+note: this linter can be disabled with `set_option linter.style.commandStart false`
+-/
+#guard_msgs in
+example {a :Nat} : a = a := rfl
```
Comment (grunweg) [inline MathlibTest/PPRoundtrip.lean:85]: Does the linter also catch the following cases/can you add those as well, please?
```
example  {a :Nat} : a = a := rfl
example {a : Nat}{b : Nat} : a = a := rfl
example {a : Nat}  : a = a := rfl
example {alpha: Type*} [Neg alpha   ] {a : Nat} : a = a := rfl

**Review by grunweg — COMMENTED** [review]
Thanks for starting this discussion! I wonder about a few more cases - can you add them as tests if they are already covered, please?

**Comment by grunweg** [comment]
Eventually, you want to import this linter in `Mathlib.Init`, right? (The current state is a decent start, it looks like there will be enough things to fix already.) How hard is it to write an auto-fixing script for the linter violations, to help with that?

**Comment by grunweg** [comment]
With the above fixes, I believe things are fine: let's see if I overlooked something when scrolling.

**Comment by leanprover-community-bot-assistant** [comment]
This pull request has conflicts, please merge `master` and resolve them.

File: Mathlib/Tactic/Linter/CommandStart.lean [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+          but all commands should start at the beginning of the line."
+  -- We only lint up to the position given by `lintUpTo`
+  if let some finalLintPos := lintUpTo stx then
+    if let some stype := stx.find? (unlintedNodes.contains ·.getKind) then
+      Linter.logLintIf linter.style.commandStart.verbose (stx.getHead?.getD stx)
+        m!"Found '{stype.getKind}' in '{stype}'"
+      if let some pos := stype.getPos? then
+        if pos.1 ≤ finalLintPos.1 then
+          return
+    let stx := capSyntax stx finalLintPos.1
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/CommandStart.lean:117]: I didn't scrutinise the tests file yet: it would be great to test this line, to clarify the behaviour. (Basically, my ideal is that if the tests pass, a linter change is correct: if mathlib exposes further edge cases, adding them as tests would be great.) With that in mind, does the file already have tests for
- unlinted nodes being left alone
- syntax before an unlinted node gets linted?

File: Mathlib/Topology/Algebra/UniformField.lean [hunk 3]
```diff
@@ -94,6 +94,7 @@ open Classical in
 The value of `hat_inv` at zero is not really specified, although it's probably zero.
 Here we explicitly enforce the `inv_zero` axiom.
 -/
+open Classical in
```
Comment (grunweg) [inline Mathlib/Topology/Algebra/UniformField.lean:97]: This was a bad merge on my part:
```suggestion

File: Mathlib/Tactic/Linter/CommandStart.lean [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      Linter.logLintIf linter.style.commandStart.verbose (stx.getHead?.getD stx)
+        m!"slightly polished source:\n'{real}'\n\n\
+          actually used source:\n'{furtherFormatting (removeComments real)}'\n\n\
+          reference formatting:\n'{st}'\n\n\
+          intermediate reference formatting:\n'{fmt}'\n"
+      if ! st.startsWith (furtherFormatting (removeComments real)) then
+        let diff := real.firstDiffPos st
+        let pos := posToShiftedPos lths diff.1 + origSubstring.startPos.1
+        let f := origSubstring.str.drop (pos)
+        let extraLth := (f.takeWhile (· != st.get diff)).length
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/CommandStart.lean:139]: This is my only major comment - and it's somewhat pre-existing from the ppRoundtrip linter (but relevant now, that I'd like this linter to be enabled by default). The simplest thing to do would be to show the old and new lines; I appreciate that the linter is not doing just that (as "eyeball this line to find the change" works surprisingly well, but can be tedious). 

The zooming in around the diff shows a short snippet of what changes (which makes the change easier to spot) --- but the strict length of just seven characters (in the simplest case) gives suggestions which are also hard to parse. `Current syntax:  'le {a: Nat'`, `Expected syntax: 'le {a : Na'` is simply not pretty.
Do you see a reasonable way to at least print "full units" (e.g., up to a corresponding closing brackets, for instance)? That's probably hard in general. Typing this now, though, perhaps a heuristic of "extend until up the corresponding opening/closing bracket" could already work well.

What do you think?

**Review by grunweg — COMMENTED** [review]
I took a first detailed look at the implementation. Here's my usual dozen of comments. Would you like some help with e.g. adding some of the tests I asked for?

**Comment by grunweg** [comment]
(I just saw that the linter is still work in progress - so you may be aware of all of these. In that case, feel free to take this as an indication "I also find that important".)

**Comment by leanprover-community-bot-assistant** [comment]
This pull request has conflicts, please merge `master` and resolve them.

**Comment by adomani** [comment]
Michael, thank you for the review!  This linter is maturing into #24465.  That linter is much more flexible and almost ready for review.  I just need to audit its exceptions and tweak it a little.

Feel free to take a look at it yourself, if you're willing to PR some of its suggestions or figure out what goes wrong.  Right now, I'm aware of poor handling of comments by the linter.

**Comment by grunweg** [comment]
That linter looks great! I'm taking a look at its current errors now, starting from the back of the output (and will push some fixes and some things I noticed there).

**Comment by adomani** [comment]
Thank you for all the comments!

For the record (and as you know), I prefer the linter at #24465, so I am closing this PR!
